Fixes #10029 - Docker registry API

daviddavis opened the pull request.

eLobato commented on app/controllers/api/v2/docker_registries_controller.rb:
```diff
@@ -0,0 +1,70 @@
+        @registry
+      end
+
+      def docker_registry_url(registry)
```
Is this method used anywhere?

daviddavis replied:
Yea, I got an undefined method without it.

daviddavis replied:
Here's the error:

```
undefined method `docker_registry_url' for #<Api::V2::DockerRegistriesController:0x0000000ced84b8> (NoMethodError)
/home/dadavis/.rvm/gems/ruby-2.0.0-p594@foreman/gems/actionpack-3.2.21/lib/action_dispatch/routing/polymorphic_routes.rb:129:in `polymorphic_url'
/home/dadavis/.rvm/gems/ruby-2.0.0-p594@foreman/gems/actionpack-3.2.21/lib/action_dispatch/routing/url_for.rb:150:in `url_for'
/home/dadavis/.rvm/gems/ruby-2.0.0-p594@foreman/gems/actionpack-3.2.21/lib/action_controller/metal/rendering.rb:60:in `_process_options'
/home/dadavis/.rvm/gems/ruby-2.0.0-p594@foreman/gems/actionpack-3.2.21/lib/action_controller/metal/streaming.rb:208:in `_process_options'
/home/dadavis/.rvm/gems/ruby-2.0.0-p594@foreman/gems/actionpack-3.2.21/lib/action_controller/metal/renderers.rb:34:in `block in _handle_render_options'
/home/dadavis/.rvm/rubies/ruby-2.0.0-p594/lib/ruby/2.0.0/set.rb:232:in `each_key'
/home/dadavis/.rvm/rubies/ruby-2.0.0-p594/lib/ruby/2.0.0/set.rb:232:in `each'
/home/dadavis/.rvm/gems/ruby-2.0.0-p594@foreman/gems/actionpack-3.2.21/lib/action_controller/metal/renderers.rb:32:in `_handle_render_options'
/home/dadavis/.rvm/gems/ruby-2.0.0-p594@foreman/gems/actionpack-3.2.21/lib/action_controller/metal/renderers.rb:28:in `render_to_body'
/home/dadavis/.rvm/gems/ruby-2.0.0-p594@foreman/gems/actionpack-3.2.21/lib/action_controller/metal/compatibility.rb:50:in `render_to_body'
/home/dadavis/.rvm/gems/ruby-2.0.0-p594@foreman/gems/actionpack-3.2.21/lib/abstract_controller/rendering.rb:88:in `render'
/home/dadavis/.rvm/gems/ruby-2.0.0-p594@foreman/gems/actionpack-3.2.21/lib/action_controller/metal/rendering.rb:16:in `render'
/home/dadavis/.rvm/gems/ruby-2.0.0-p594@foreman/gems/actionpack-3.2.21/lib/action_controller/metal/instrumentation.rb:40:in `block (2 levels) in render'
/home/dadavis/.rvm/gems/ruby-2.0.0-p594@foreman/gems/activesupport-3.2.21/lib/active_support/core_ext/benchmark.rb:5:in `block in ms'
/home/dadavis/.rvm/rubies/ruby-2.0.0-p594/lib/ruby/2.0.0/benchmark.rb:296:in `realtime'
/home/dadavis/.rvm/gems/ruby-2.0.0-p594@foreman/gems/activesupport-3.2.21/lib/active_support/core_ext/benchmark.rb:5:in `ms'
/home/dadavis/.rvm/gems/ruby-2.0.0-p594@foreman/gems/actionpack-3.2.21/lib/action_controller/metal/instrumentation.rb:40:in `block in render'
/home/dadavis/.rvm/gems/ruby-2.0.0-p594@foreman/gems/actionpack-3.2.21/lib/action_controller/metal/instrumentation.rb:83:in `cleanup_view_runtime'
/home/dadavis/.rvm/gems/ruby-2.0.0-p594@foreman/gems/activerecord-3.2.21/lib/active_record/railties/controller_runtime.rb:24:in `cleanup_view_runtime'
/home/dadavis/.rvm/gems/ruby-2.0.0-p594@foreman/gems/actionpack-3.2.21/lib/action_controller/metal/instrumentation.rb:39:in `render'
/home/dadavis/.rvm/gems/ruby-2.0.0-p594@foreman/gems/actionpack-3.2.21/lib/action_controller/metal/responder.rb:254:in `display'
/home/dadavis/.rvm/gems/ruby-2.0.0-p594@foreman/gems/actionpack-3.2.21/lib/action_controller/metal/responder.rb:204:in `api_behavior'
/home/dadavis/Projects/foreman/app/controllers/api/api_responder.rb:12:in `api_behavior'
/home/dadavis/.rvm/gems/ruby-2.0.0-p594@foreman/gems/actionpack-3.2.21/lib/action_controller/metal/responder.rb:181:in `rescue in to_format'
/home/dadavis/.rvm/gems/ruby-2.0.0-p594@foreman/gems/actionpack-3.2.21/lib/action_controller/metal/responder.rb:175:in `to_format'
/home/dadavis/.rvm/gems/ruby-2.0.0-p594@foreman/gems/actionpack-3.2.21/lib/action_controller/metal/responder.rb:153:in `respond'
/home/dadavis/.rvm/gems/ruby-2.0.0-p594@foreman/gems/actionpack-3.2.21/lib/action_controller/metal/responder.rb:146:in `call'
/home/dadavis/.rvm/gems/ruby-2.0.0-p594@foreman/gems/actionpack-3.2.21/lib/action_controller/metal/mime_responds.rb:239:in `respond_with'
/home/dadavis/Projects/foreman/app/controllers/api/base_controller.rb:94:in `process_success'
/home/dadavis/Projects/foreman/app/controllers/api/base_controller.rb:99:in `process_response'
/home/dadavis/Projects/foreman-docker/app/controllers/api/v2/docker_registries_controller.rb:39:in `create'
/home/dadavis/.rvm/gems/ruby-2.0.0-p594@foreman/gems/actionpack-3.2.21/lib/action_controller/metal/implicit_render.rb:4:in `send_action'
/home/dadavis/.rvm/gems/ruby-2.0.0-p594@foreman/gems/actionpack-3.2.21/lib/abstract_controller/base.rb:167:in `process_action'
/home/dadavis/.rvm/gems/ruby-2.0.0-p594@foreman/gems/actionpack-3.2.21/lib/action_controller/metal/rendering.rb:10:in `process_action'
/home/dadavis/.rvm/gems/ruby-2.0.0-p594@foreman/gems/actionpack-3.2.21/lib/abstract_controller/callbacks.rb:18:in `block in process_action'
/home/dadavis/.rvm/gems/ruby-2.0.0-p594@foreman/gems/activesupport-3.2.21/lib/active_support/callbacks.rb:540:in `block (5 levels) in _run__4345451274361731940__process_action__892029585679967221__cal
lbacks'
/home/dadavis/.rvm/gems/ruby-2.0.0-p594@foreman/gems/activesupport-3.2.21/lib/active_support/callbacks.rb:215:in `block in _conditional_callback_around_20290'
/home/dadavis/Projects/foreman/app/controllers/api/v2/base_controller.rb:151:in `disable_json_root'
/home/dadavis/.rvm/gems/ruby-2.0.0-p594@foreman/gems/activesupport-3.2.21/lib/active_support/callbacks.rb:214:in `_conditional_callback_around_20290'
/home/dadavis/.rvm/gems/ruby-2.0.0-p594@foreman/gems/activesupport-3.2.21/lib/active_support/callbacks.rb:539:in `block (4 levels) in _run__4345451274361731940__process_action__892029585679967221__cal
lbacks'
/home/dadavis/.rvm/gems/ruby-2.0.0-p594@foreman/gems/activesupport-3.2.21/lib/active_support/callbacks.rb:215:in `block in _conditional_callback_around_20289'
/home/dadavis/.rvm/gems/ruby-2.0.0-p594@foreman/gems/activesupport-3.2.21/lib/active_support/callbacks.rb:326:in `around'
/home/dadavis/.rvm/gems/ruby-2.0.0-p594@foreman/gems/activesupport-3.2.21/lib/active_support/callbacks.rb:310:in `_callback_around_19749'
/home/dadavis/.rvm/gems/ruby-2.0.0-p594@foreman/gems/activesupport-3.2.21/lib/active_support/callbacks.rb:214:in `_conditional_callback_around_20289'
/home/dadavis/.rvm/gems/ruby-2.0.0-p594@foreman/gems/activesupport-3.2.21/lib/active_support/callbacks.rb:516:in `block (3 levels) in _run__4345451274361731940__process_action__892029585679967221__cal
lbacks'
/home/dadavis/.rvm/gems/ruby-2.0.0-p594@foreman/gems/activesupport-3.2.21/lib/active_support/callbacks.rb:215:in `block in _conditional_callback_around_20288'
```

eLobato replied:
Oh I see. This is not very relevant, it's just the ApiResponder trying to get the URL of registries, which is actually `registry_url` - or http://test.host/registries/1 - not the docker registry URL. If you rename this as simply 'registry' instead of 'docker_registries' you can get rid of this method easily, and the '/**docker**/api/v2/**docker**_registries/` duplicity.

eLobato commented on test/functionals/api/v2/docker_registries_controller_test.rb:
```diff
@@ -0,0 +1,55 @@
+      end
+
+      test 'update a docker registry' do
+        registry = FactoryGirl.create(:docker_registry)
```
All this actions that require a registry perhaps could be put in a context with a setup block that initializes the registry.

daviddavis replied:
Will do.

eLobato commented on config/routes.rb:
```diff
@@ -38,6 +38,7 @@
             end
           end
         end
+        resources :docker_registries, :except => [:new, :edit]
```
Since the API is namespaced, (`docker/api/v2/`) does saying again `*docker*_registries` serves any purpose vs just `registries`?

daviddavis replied:
Originally, I did registries but I couldn't get the action permission tests to pass. If you know how to do that, then let me know.

eLobato commented on app/controllers/api/v2/docker_registries_controller.rb:
```diff
@@ -0,0 +1,70 @@
+        process_response @docker_registry.destroy
+      end
+
+      # rubocop:disable Style/TrivialAccessors, Style/AccessorMethodName
```
Why is this disabled here ?

daviddavis replied:
I'm overriding a method in base controller. I don't want to have to rewrite all the api code to fix this rubocop error.

daviddavis replied:
Here's the method:

https://github.com/theforeman/foreman/blob/develop/app/controllers/api/base_controller.rb#L34-L36

Here's why it doesn't cause a rubocop error in foreman:

https://github.com/theforeman/foreman/blob/develop/.rubocop_todo.yml#L458-L462
https://github.com/theforeman/foreman/blob/develop/.rubocop_todo.yml#L112-L114

eLobato replied:
But that's only there because the variable goes from @docker_registry to @registry above (I can't see a reason for this other than saving characters?). This passes tests after you remove `get_resource`:

```ruby
def show
end

def create 
 @docker_registry = DockerRegistry.new(params[:docker_registry])
 process_response @docker_registry.save 
end

def update
  process_response @docker_registry.update_attributes(params[:docker_registry])
end
```

daviddavis replied:
I did it to remain consistent with the existing registry controller and views:

https://github.com/theforeman/foreman-docker/blob/master/app/controllers/registries_controller.rb
https://github.com/theforeman/foreman-docker/tree/master/app/views/registries

eLobato commented:
@daviddavis Great work! It works just as expected :clap: . There are a few minor warts we can probably iron out easily, I left some comments inline, let me know your thoughts about them please.

Thanks again!

daviddavis commented:
@eLobato Updated the test. If you have any idea how to change the registries route and get the perms working, let me know.

eLobato commented:
@daviddavis I submitted a PR to your fork, it changes the route, access_permissions_test works, there's a minor thing I haven't fixed on update. https://github.com/daviddavis/foreman-docker/pull/2

daviddavis commented:
I think defining `resource_class` was the piece I was missing when I tried to use `/docker/api/v2/registries`. I've merged your PR, fixed the update test, and updated this PR.

eLobato commented:
Merged as a29cf1c2f7466ef3a0ffb0fb11b5fca00fa6635c, thanks @daviddavis !
